Added Support for a UI display option for how to format fields marked as Offsets/Sizes

Zorro666 opened the pull request:
## Description

Issue [2924](https://github.com/baldurk/renderdoc/issues/2924)

Previously by default, values are displayed using decimal formatting unless they are above a threshold (0xFFFFFF) then they are displayed using hexadecimal formatting.

There is now a UI formatter setting to change this behaviour for fields which have been marked as an Offset/Size type. 
The UI option has three modes: `Auto`, `Decimal`, `Hexadecimal`: 
- `Auto` : behaves as it currently does, decimal by default with hexadecimal when the value is above a threshold
- `Decimal` : always display the value using decimal formatting
- `Hexadecimal` : always display the value using hexadecimal formatting

The new UI option does not affect fields which have **not** been marked as an Offset/Size type.

There is one additional change to existing behaviour when printing using hexadecimal (`Auto` or `Hexadecimal` mode) then for values below `UINT32_MAX` the values are treated as `uint32_t` instead of `uint64_t` (eight leading zero's instead of sixteen leading zero's) e.g.
- Before
`dstOffset 0x0000000001000000`
- After
`dstOffset 0x01000000`

This is an example of a field not marked as an Offset or Size field with a large value, the display does not change for `Auto`, `Decimal`, `Hexadecimal` mode:
- Before
`jakeTest 0x0000000001000000`
- After
`jakeTest 0x01000000`

Done a first pass, reviewing the serialised fields for all APIs to apply mark up for the `OffsetOrSize` field type. The selection criteria was an offset or size which is measured in bytes. This markup can be refined in the future it does not affect the captured data.

## Testing

### Before showing the existing behaviour of the `Auto` formatting mode
- Decimal for values less than or equal to 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/b5f3c0cc-4e1f-417d-b267-fd4268ca10d5)- Hexadecimal for values above 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/e1b52e3f-8935-477b-a1fb-166d2babd393)

### After showing the behaviour of the new formatting modes

#### `Auto` formatting mode
- Decimal for values less than or equal to 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/1d86b856-2549-4930-b73a-878ba206cb9f)- Hexadecimal for values above 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/012ef451-7002-4f26-b091-38f6b44f97c4)

#### `Decimal` formatting mode
- Decimal for values less than or equal to 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/776ceb70-77d5-4a9c-92b4-78e562d4e117)- Decimal for values above 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/f83a7c7b-e7d6-4421-b1c8-c851e8c8969d)

#### `Hexadecimal` formatting mode
- Hexadecimal for values less than or equal to 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/eb0df7fc-7bc2-4c71-910a-6140a02173cb)- Hexadecimal for values above 0xFFFFFF
![image](https://github.com/Zorro666/renderdoc/assets/39392/63a475e4-4dbb-4e3d-9b9e-85d4d66581e0)

Zorro666 committed:
UI and Config support for dec/hex display option for Offsets or Sizes

Used when fields are marked as being an Offset or Size
Include simple latch to trigger a UI refresh when closing the settings dialog.
Currently only connected to if the Offset/Size format option is altered

Zorro666 committed:
Sparse structs serialisation markup for Offset, Size type fields

Zorro666 committed:
D3D11 serialisation markup for Offset, Size type fields

Zorro666 committed:
D3D12 serialisation markup for Offset, Size type fields

Zorro666 committed:
GL serialisation markup for Offset, Size type fields

Zorro666 committed:
Vulkan serialisation markup for Offset, Size type fields

Zorro666 commented on qrenderdoc/Code/Interface/PersistantConfig.h:
```diff
@@ -372,6 +372,16 @@ DECLARE_REFLECTION_STRUCT(BugReport);
+  DOCUMENT(                                                                                        \
+      "The formatting mode to use for values marked as Offsets or Sizes.\n"                        \
+      "\n"                                                                                         \
+      "E.g. Auto: decimal by default and hexadecimal if above a certain threshold, "               \
```
I was not sure about the formatting style to use here.
If I build documentation locally will I see these changes?

baldurk replied:
Yes these will be documented on the `PersistantConfig`. This formatting seems fine to me.

Zorro666 commented on qrenderdoc/Windows/Dialogs/SettingsDialog.ui:
```diff
@@ -287,6 +287,23 @@ e.g. 1000 * 10 = 1e4</string>
+             <string>Formatting mode to use for fields marked as a byte offset or a byte size</string>
+            </property>
+            <property name="text">
+             <string>Offset or size fields format mode</string>
```
I couldn't decide if this should be
"Offset or size"
or
"Offset or Size"

else where have been referring to this field type as Offset or Size

baldurk replied:
I think either is probably fine, I slightly prefer this way.

Zorro666 commented on renderdoc/core/sparse_page_table.cpp:
```diff
@@ -913,7 +913,7 @@ template <typename SerialiserType>
 void DoSerialise(SerialiserType &ser, Sparse::Page &el)
```
Was not sure about marking up helper structures but it seemed to fit the criteria of an offset or size measured in bytes

baldurk replied:
It's probably better to err on the side of caution and annotate these just in case, but really we should be hiding these structures & chunks entirely from the user. Anything that doesn't come from the API itself is liable to cause confusion.

baldurk reviewed (changes requested):
This looks good to me. In terms of annotating serialised data I've checked that nothing is wrongly marked up, it will be easier to look for missing annotations once this has been merged and I can search in the code directly.

Have you looked at the offsets or sizes displayed in the pipeline state, or in the buffer viewer? There are some places where sizes are displayed without coming directly from serialised data.

Zorro666 committed:
Offset and Size consistency display in BufferViewer UI

Use Formatter::HumanFormat() to get consistent formatting and to apply the UI formatting configuration mode for Offsets and Sizes i.e. Auto, Decimal, Hexadecimal.

Zorro666 committed:
Offset and Size consistency display in PipelineState UI

Use Formatter::HumanFormat() to get consistent formatting and to apply the UI formatting configuration mode for Offsets and Sizes i.e. Auto, Decimal, Hexadecimal.

Zorro666 committed:
Vulkan Mesh Shader serialisation markup for Offset, Size type fields
